include/revng/ADT/GenericGraph.h

```diff
@@ -353,11 +370,557 @@ class BidirectionalNode
+      Neighbor(E.Neighbor), Label(E.Label.get()) {}
+    explicit EdgeView(NonOwningEdge<NodeType, LabelType> &E) :
+      Neighbor(E.Neighbor), Label(E.Label) {}
+  };
+
+  template<typename NodeType, typename LabelType>
+  struct ConstEdgeView {
+    NodeType const *Neighbor;
+    LabelType const *Label;
+
+    explicit ConstEdgeView(OwningEdge<NodeType, LabelType> const &E) :
+      Neighbor(E.Neighbor), Label(E.Label.get()) {}
+    explicit ConstEdgeView(NonOwningEdge<NodeType, LabelType> const &E) :
+      Neighbor(E.Neighbor), Label(E.Label) {}
+  };
+
+  template<typename NodeType>
+  struct Unlabeled {
+    NodeType *Neighbor;
+  };
+
+  template<typename NodeType>
+  struct UnlabeledView {
+    NodeType &Neighbor;
+
+    explicit UnlabeledView(Unlabeled<NodeType> &E) : Neighbor(*E.Neighbor) {}
+  };
+
+  template<typename NodeType>
+  struct ConstUnlabeledView {
+    NodeType const &Neighbor;
+
+    explicit ConstUnlabeledView(Unlabeled<NodeType> const &E) :
+      Neighbor(*E.Neighbor) {}
+  };
+} // namespace detail
+
+/// A node type with support for non-trivial edge types.
+template<typename Node,
+         typename EdgeLabel = Empty,
+         bool HasParent = true,
+         size_t SmallSize = 2,
+         typename FinalType = std::false_type,
+         size_t ParentSmallSize = 16,
+         bool ParentHasEntryNode = true>
+class MutableEdgeNode
+  : public detail::MutableEdgeNodeBaseTCalc<Node,
+                                            EdgeLabel,
+                                            HasParent,
+                                            SmallSize,
+                                            MutableEdgeNode,
+                                            FinalType,
+                                            ParentSmallSize,
+                                            ParentHasEntryNode>::Result {
+public:
+  static constexpr bool is_mutable_edge_node = true;
+  static constexpr bool has_parent = HasParent;
+  using TypeCalc = detail::MutableEdgeNodeBaseTCalc<Node,
+                                                    EdgeLabel,
+                                                    HasParent,
+                                                    SmallSize,
+                                                    MutableEdgeNode,
+                                                    FinalType,
+                                                    ParentSmallSize,
+                                                    ParentHasEntryNode>;
+  using DerivedType = typename TypeCalc::DerivedType;
+  using Base = typename TypeCalc::Result;
+
+  static constexpr bool AreEdgesLabeled = !std::is_same_v<EdgeLabel, Empty>;
+
+public:
+  using Edge = EdgeLabel;
+  using EdgeView = std::conditional_t<AreEdgesLabeled,
+                                      detail::EdgeView<DerivedType, EdgeLabel>,
+                                      detail::UnlabeledView<DerivedType>>;
+  using ConstEdgeView = std::conditional_t<
```

Comment:
Yeah, I've lately been having some weird problems with our `clang-format`. Sometimes it indents some parts, then it wants them to be changed back.

And I have no idea what to even blame it on.

Even now, I rerun it and it attempts to undo a couple of changes I literally made just because it complained (`Fix whitespace problems founds after the update of clang-format` commit).

Reply:
`clang-format` does some stuff that might not be compatible with other checks performed by `check-revng-conventions`.
This often leads to reformulate the code splitting it in multiple lines.

Reply:
I manually went through the `rebase` running `./scripts/check-revng-conventions --force-format` on each commit, so this should be fine now.